src/main/java/com/avaya/letscode/Cart.java

```diff
@@ -7,55 +7,61 @@
 
 public class Cart {
     //List of items in the cart with respective unit quantity
-    Map<Product, CartItem> items;       
+    private final Map<Product, CartItem> items;
     
     //List of offers applicable to the cart
     // Map<String, Offer> cartOffers;
-    List<Offer> cartOffers;
-
+    private List<Offer> cartOffers;
 
     int quantity = 0;       //Total number of items in cart
-    public boolean isEmpty() {
-        return quantity == 0;
-    }
 
     Cart(){
         this.items = new HashMap<>();
         this.cartOffers = new ArrayList<>();
     }
+
+    public boolean isEmpty() {
+        return quantity == 0;
+    }
+
     public void AddItem(Product product) {
         this.AddItem(product, 1);
     }
 
-    public int getQuantity() {
-        return quantity;
+    public void AddItem(Product product, int quantity) {
+        this.AddItem(product, quantity,"");
     }
 
-    public double getCartValue() {
+    public void AddItem(Product product, int quantity, String offer) {
+        this.quantity += quantity;
+        CartItem cartItem = new CartItem(product, quantity, Offer.create(offer));
+        this.items.merge(product, cartItem, CartItem::merge);
+    }
 
-        double cartValue = this.items.values().stream().mapToInt(CartItem::getCartItemValue).sum();
-        //stream offer list and apply each offer to the cart
-        double discount = this.cartOffers.stream().mapToDouble(offer -> offer.getDiscountPercentage(this)).sum();
-        return cartValue - cartValue*discount/100;
-        
+    public void AddOffer(String offer) {
+        this.cartOffers.add(Offer.create(offer));
     }
 
-    public void AddItem(Product product, int quantity) {
-        this.AddItem(product, quantity,"");
+    public int getQuantity() {
+        return quantity;
     }
 
     public int getQuantity(Product product) {
         return this.items.containsKey(product) ? this.items.get(product).getQuantity() : 0;
     }
 
-    public void AddItem(Product product, int quantity, String offer) {
-        this.quantity += quantity;
+    public double getCartValue() {
 
-        CartItem cartItem = new CartItem(product, quantity, Offer.create(offer));
-        this.items.merge(product, cartItem, CartItem::merge);
+        //stream offer list and apply each offer to the cart and find the minimum value
+        // if cart has no offers, return the cart value before offer
+        if (this.cartOffers.isEmpty()) 
+            return this.getCartValueBeforeOffer();
+        else
+            return this.cartOffers.stream().mapToDouble(offer -> offer.getDiscountedValue(this)).min().getAsDouble();
```

Comment:
This code is not covered in any of the tests. May be we should add a test for the scenario of - applying the best offer when multiple offers are added.

Reply:
Done